internal/kms/kms.go

```diff
@@ -216,6 +221,344 @@ func (k *Kms) CreateKeys(ctx context.Context, scopeId string, opt ...Option) err
+
+	opts := getOpts(opt...)
+	kmsOpts := []wrappingKms.Option{
+		wrappingKms.WithRandomReader(opts.withRandomReader),
+		wrappingKms.WithTx(opts.withTx),
+		wrappingKms.WithRewrap(opts.withRewrap),
+	}
+
+	switch {
+	case !isNil(opts.withReader) && isNil(opts.withWriter):
+		return errors.New(ctx, errors.InvalidParameter, op, "missing writer")
+	case isNil(opts.withReader) && !isNil(opts.withWriter):
+		return errors.New(ctx, errors.InvalidParameter, op, "missing reader")
+	case !isNil(opts.withReader) && !isNil(opts.withWriter):
+		r, ok := opts.withReader.(*db.Db)
+		if !ok {
+			return errors.New(ctx, errors.InvalidParameter, op, "unable to convert reader to db.Db")
+		}
+		w, ok := opts.withWriter.(*db.Db)
+		if !ok {
+			return errors.New(ctx, errors.InvalidParameter, op, "unable to convert writer to db.Db")
+		}
+		kmsOpts = append(kmsOpts, wrappingKms.WithReaderWriter(db.NewChangeSafeDbwReader(r), db.NewChangeSafeDbwWriter(w)))
+	}
+
+	err := k.underlying.RotateKeys(ctx, scopeId, kmsOpts...)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	return nil
+}
+
+// ListDataKeyVersionReferencers will list the names of all tables
+// referencing the private_id column of the data key version table.
+// Supported options:
+//   - WithTx
+//   - WithReader (requires WithWriter)
+//   - WithWriter (requires WithReader)
+func (k *Kms) ListDataKeyVersionReferencers(ctx context.Context, opt ...Option) ([]string, error) {
+	const op = "kms.(Kms).ListDataKeyVersionReferencers"
+
+	opts := getOpts(opt...)
+	kmsOpts := []wrappingKms.Option{
+		wrappingKms.WithTx(opts.withTx),
+	}
+
+	switch {
+	case !isNil(opts.withReader) && isNil(opts.withWriter):
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing writer")
+	case isNil(opts.withReader) && !isNil(opts.withWriter):
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing reader")
+	case !isNil(opts.withReader) && !isNil(opts.withWriter):
+		r, ok := opts.withReader.(*db.Db)
+		if !ok {
+			return nil, errors.New(ctx, errors.InvalidParameter, op, "unable to convert reader to db.Db")
+		}
+		w, ok := opts.withWriter.(*db.Db)
+		if !ok {
+			return nil, errors.New(ctx, errors.InvalidParameter, op, "unable to convert writer to db.Db")
+		}
+		kmsOpts = append(kmsOpts, wrappingKms.WithReaderWriter(db.NewChangeSafeDbwReader(r), db.NewChangeSafeDbwWriter(w)))
+	}
+
+	refs, err := k.underlying.ListDataKeyVersionReferencers(ctx, kmsOpts...)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	return refs, nil
+}
+
+// ListDataKeyVersionDestructionJobs lists any in-progress data key destruction jobs in the scope.
+// Options are ignored.
+func (k *Kms) ListDataKeyVersionDestructionJobs(ctx context.Context, scopeId string, _ ...Option) ([]*DataKeyVersionDestructionJobProgress, error) {
+	var jobs []*DataKeyVersionDestructionJobProgress
+	if err := k.reader.SearchWhere(ctx, &jobs, "scope_id=?", []interface{}{scopeId}); err != nil {
+		return nil, err
```

Comment:
should this err be wrapped?

Reply:
it would seem that this error path would be fairly simple to test.

Reply:
I added a wrapping to this error, good shout.

I don't think this error path is so easy to test, if the scope ID doesn't exist it simply returns no jobs. Maybe I'm missing something?

Reply:
why not write a test that gives the kms a db without the scope_id for a unit test?